core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/uni/IfExistsUniWithUniNode.java

```diff
@@ -0,0 +1,243 @@
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.optaplanner.constraint.streams.bavet.bi.JoinBiNode;
+import org.optaplanner.constraint.streams.bavet.common.AbstractNode;
+import org.optaplanner.constraint.streams.bavet.common.BavetTupleState;
+import org.optaplanner.constraint.streams.bavet.common.index.Indexer;
+
+public final class IfExistsUniWithUniNode<A, B> extends AbstractNode {
+
+    private final boolean shouldExist;
+    private final Function<A, Object[]> mappingA;
+    private final Function<B, Object[]> mappingB;
+    private final int inputStoreIndexA;
+    private final int inputStoreIndexB;
+    /**
+     * Calls for example {@link UniScorer#insert(UniTuple)}, {@link JoinBiNode#insertA(UniTuple)} and/or ...
+     */
+    private final Consumer<UniTuple<A>> nextNodesInsert;
+    /**
+     * Calls for example {@link UniScorer#retract(UniTuple)}, {@link JoinBiNode#retractA(UniTuple)} and/or ...
+     */
+    private final Consumer<UniTuple<A>> nextNodesRetract;
+    // No outputStoreSize because this node is not a tuple source, even though it has a dirtyCounterQueue.
+
+    private final Indexer<UniTuple<A>, Counter<A>> indexerA;
+    private final Indexer<UniTuple<B>, Set<Counter<A>>> indexerB;
+    private final Queue<Counter<A>> dirtyCounterQueue;
+
+    public IfExistsUniWithUniNode(boolean shouldExist, Function<A, Object[]> mappingA, Function<B, Object[]> mappingB,
+            int inputStoreIndexA, int inputStoreIndexB,
+            Consumer<UniTuple<A>> nextNodesInsert, Consumer<UniTuple<A>> nextNodesRetract,
+            Indexer<UniTuple<A>, Counter<A>> indexerA, Indexer<UniTuple<B>, Set<Counter<A>>> indexerB) {
+        this.shouldExist = shouldExist;
+        this.mappingA = mappingA;
+        this.mappingB = mappingB;
+        this.inputStoreIndexA = inputStoreIndexA;
+        this.inputStoreIndexB = inputStoreIndexB;
+        this.nextNodesInsert = nextNodesInsert;
+        this.nextNodesRetract = nextNodesRetract;
+        this.indexerA = indexerA;
+        this.indexerB = indexerB;
+        dirtyCounterQueue = new ArrayDeque<>(1000);
+    }
+
+    public void insertA(UniTuple<A> tupleA) {
+        if (tupleA.store[inputStoreIndexA] != null) {
+            throw new IllegalStateException("Impossible state: the input for the fact ("
+                    + tupleA.factA + ") was already added in the tupleStore.");
+        }
+        Object[] indexProperties = mappingA.apply(tupleA.factA);
+        tupleA.store[inputStoreIndexA] = indexProperties;
+
+        Counter<A> counter = new Counter<>(tupleA);
+        indexerA.put(indexProperties, tupleA, counter);
+
+        Map<UniTuple<B>, Set<Counter<A>>> counterSetMapB = indexerB.get(indexProperties);
+        counter.countB = counterSetMapB.size();
+        if (shouldExist ? counter.countB > 0 : counter.countB == 0) {
```

Comment:
This is a confusing expression. Adding parens around the if/else branches would make it a bit more self-explanatory, I think.

Reply:
I agree, I 'll do that once I start supporting ifNotExists() decently